### DISCUSSION TITLE
Autodiscovery Feedback

### DISCUSSION THREAD

**Discussion by olblak** [updatecli/updatecli#874] [Ideas]
I'll use this discussion to collect my experience after several weeks of using the the autodiscovery

**Local Mode**
I find it a bit annoying to have to right `updatecli diff --experimental --local-autodiscovery`.
I am considering shortening to `updatecli diff --experimental --local`
so replacing `--local-autodiscovery` by `local`

**Declarative**
At the moment the declarative enable all crawlers by default, and I end up disabling the crawlers that I don't want to automate.
So  I am considering only loading crawler specified in the manifest

**Comment by dduportal** [comment]
>  So I am considering only loading crawler specified in the manifest

Excellent idea!

> `--local`

My 2 cents: Saying "run an updatecli diff locally" makes no sense to me :( I find it confusing to understand "what is it going to do".

**Reply by olblak** [reply]
Initially I didn't implemented the `--local-autodiscovery` but `--disable-autodiscovery`
Local autodiscovery was enabled by default but it introduced a lot of of noise.
Because running `updatecli diff` would run pipelines specified in `updatecli.yaml`, `updatecli.d`, and the "local" autodiscovery.
I had a lot of duplication as a pipeline could come from a regular manifest, an autodiscovery manifest,  and or the default local autodiscovery.

I couldn't think about a way to solve this without a major refactoring, hence why I introduce the `--local-autodiscovery`

The refactoring that I have in my head is, instead of running sequentially each pipeline one by one, we would build a tree of "action" so each action would be executed once.

Differently said, if we have three Updatecli manifest that specify the same spec for the same source "kind", in the current state we run it three times instead of running it once and reusing the same result for the other pipeline.

Even thought I don't think that this refactoring will be difficult, I think we have other more interesting things to improve

**Reply by olblak** [reply]
Another option would be 

`updatecli diff` do not load "updatecli.d" and "updatecli.yaml" by default anymore but only the default autodiscovery

**Reply by olblak** [reply]
I have been thinking about this for a while. While I don't really found the best solution, I think the worst least, is to fallback to local-discovery if no configuration is available.

So `updatecli diff` should triggerer --local-autodiscovery if we don't have a file named "updatecli.yaml" or "updatecli.d"

**Reply by dduportal** [reply]
Make sense. Don't forget to add an "INFO" message to end user even by default

**Reply by olblak** [reply]
I am a bit annoyed by the behavior.

```
+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "updatecli/updatecli.d/golang.yaml"
Loading Pipeline "updatecli/updatecli.d/golangci-lint.yaml"
Loading Pipeline "updatecli/updatecli.d/updatecli.yaml"
Loading Pipeline "updatecli/updatecli.d/venom.yaml"
ERROR: failed loading pipeline(s)
	* "updatecli/updatecli.d/golang.yaml" - updatecli version compatibility error:
pipeline "Bump Golang Version" requires Updatecli version greater than "0.34.0", skipping
	* "updatecli/updatecli.d/golangci-lint.yaml" - updatecli version compatibility error:
pipeline "Bump golangci-lint version" requires Updatecli version greater than "0.34.0", skipping
	* "updatecli/updatecli.d/updatecli.yaml" - updatecli version compatibility error:
pipeline "Bump udpatecli version" requires Updatecli version greater than "0.34.0", skipping
	* "updatecli/updatecli.d/venom.yaml" - updatecli version compatibility error:
pipeline "Bump venom version" requires Updatecli version greater than "0.34.0", skipping

0 pipeline(s) successfully loaded

SCM repository retrieved: 0


WARNING: Default Autodiscovery crawlers has been enabled as zero Updatecli configuration were specified and we didn't detected default configuration.
```

**Comment by olblak** [comment]
First https://github.com/updatecli/updatecli/pull/877 to stop enabling all crawlers when using from a manifest

**Reply by olblak** [reply]
I opened this PR https://github.com/updatecli/updatecli/pull/965

**Comment by olblak** [comment]
So I reverted the last PR because it was really confusing when to apply manifest and when not.
In the end I think the apply manifest and autodiscovery are two differents behaviors and therefor should be call in different ways.

As we already have `updatecli manifest upgrade` and `updatecli manifest show`.
I suggest to add `updatecli manifest apply`

So the command `updatecli apply` would become `updatecli manifest apply`
and the command `updatecli diff` would become `updatecli manifest apply --dry-run`

Then we could have `updatecli auto` which wouldn't load any manifest , just default autodiscovery behavior and in a second iteration once the scm refactoring is done we could even imagine detecting scm automatically
